controller/controller_server/utils.py

```diff
@@ -8,43 +10,147 @@
 import controller.controller_server.messages as messages
 from controller.array_action.config import FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE
 from controller.common.csi_logger import get_stdout_logger
+from controller.controller_server.controller_types import ArrayConnectionInfo, ObjectIdInfo, ObjectParameters
 from controller.controller_server.errors import ObjectIdError, ValidationException
 from controller.csi_general import csi_pb2
 
 logger = get_stdout_logger()
 
 
-def get_array_connection_info_from_secret(secrets):
+def _is_topology_match(secret_topologies, node_topologies):
+    for topologies in secret_topologies:
+        logger.debug(
+            "Comparing topologies: volume topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if topologies.items() <= node_topologies.items():
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_secret_by_topologies(raw_secret_config, node_topologies):
+    secret_config = ast.literal_eval(raw_secret_config)
```

Comment:
consider wrapping this line with try-except, for a user friendly parsing error message, and moving it all to a separate function

Reply:
done